Add some initial tests as per #147

lamplightdev opened the pull request:
Hi Cory - I've had a play with the Stencil test runner and it seems to work pretty well (uses Puppeteer under the hood). I've added some tests here for the `slTabGroup`, `slDialog` and `slDrawer` components - mainly checking the basic moving parts behave as expected, that the component methods work, and that the events are firing at the right times.

There are two tests here that fail (on both `slDialog` and `slDrawer`, the `slHide` event is called twice when hiding) - I'll file a bug report.

I'm happy to continue working through the rest of the components adding similar tests if you think it's worthwhile. The test runner does have some basic support for comparing screenshots that might be useful for the visual regression tests, although it looks a bit experimental at the moment.

lamplightdev committed:
Adds stencil test dependencies

lamplightdev committed:
Add e2e tests for slTabGroup

lamplightdev committed:
Add e2e tests for slDialog

lamplightdev committed:
Add e2e tests for slDrawer

claviska commented:
Thanks, this looks awesome. Let me play with it today or tomorrow and I'll provide some feedback! (I haven't had a chance to use the built-in test runner yet.)

claviska commented:
These are great — the syntax is easy enough and the tests have already helped me uncovered a bug. 😆 

Let's roll with it!

lamplightdev committed:
Add e2e tests for slAlert

lamplightdev committed:
Add e2e tests for slButton

lamplightdev committed:
Add e2e tests for slCheckbox

lamplightdev committed:
Add e2e tests for slDetails

lamplightdev committed:
Add e2e tests for slDropdown

lamplightdev committed:
Replace flakey wait timeouts

Instead wait for `transitionend` on appropriate elements

lamplightdev committed:
Add e2e tests for slForm

lamplightdev committed:
Add e2e tests for slInput

lamplightdev committed:
Fix slFocus/slBlur transposition

lamplightdev committed:
Add e2e tests for slMenu

lamplightdev committed:
Add e2e tests for slMenuItem

lamplightdev committed:
Add e2e tests for slRadio

lamplightdev committed:
Add e2e tests for slRange

lamplightdev committed:
Add e2e tests for slRating

lamplightdev committed:
Add e2e tests for slSwitch

lamplightdev committed:
slCheckbox - add additional tests for `checked` property

lamplightdev committed:
slRadio - add additional tests for `checked` property

lamplightdev committed:
Add e2e tests for slTextarea

lamplightdev committed:
Add e2e tests for slTag

lamplightdev committed:
Add e2e tests for slSelect

lamplightdev committed:
Add value property checks for Input/Textarea

lamplightdev committed:
Remove `only` left inadvertently in tests

lamplightdev committed:
Run tests in series

Running tests in parallel can lead to timeout errors - presumably because the Puppeteer Chrome instances are heavyweight and causing slowdowns?

lamplightdev committed:
Use more reliable wait for `slAfter*` events

Instead of waiting for `transitionend` events which can be brittle, instead wait for the event itself to happen before testing for it.

lamplightdev committed:
Add e2e tests for slColorPicker

lamplightdev committed:
Run tests in parallel

Switch back to running tests in parallel as the tests are now more reliable under load.

lamplightdev commented:
I think I'm finished with this first round of e2e tests. They should cover the basic checks for all events and methods that are currently associated with any component.

I've made an effort to make them robust (avoiding any generic waits that could prove flakey) so I'm pretty confident they could be used in any CI/CD pipeline if required.

Currently there are two failing tests (reported as bug #195).

claviska commented:
Thanks again for tackling this!

How do you feel about moving each test into the same folder as its respective component? I feel like people (including myself) will be more apt to write/update tests if they're not disconnected from the source.

Do you know why some of the tests take so long? Not sure if this is normal Puppeteer behavior or if there's something we can do to speed it up.

![image](https://user-images.githubusercontent.com/55639/92283724-81686e80-eece-11ea-8963-460f5d01ad38.png)

lamplightdev commented:
With regard the time it takes to run these tests, I experience similar timings here. It seems that when they run in parallel (currently 8 max-workers) then that can max out the CPU which makes each test slower to run. If you switch to running the tests in series then each test is quicker, but the overall time is longer 😄. The browser window is set up and torn down for each test, so I imagine there's some overhead there too. If we look at the average time for a test - 22/119 ~ 0.2s / test, then considering that some of the tests have to wait for the `slAfter*` events which are only fired once the CSS transitions are finished, then I'm not sure this is too bad.

Yep, agree the test file should live with the other component files - I will make that change, and see if I can come up with a simple test for each of the components that don't currently have any.

claviska commented:
Thanks for this! A lot has changed since this PR was submitted, so I merged it and update the tests to reflect that. I'll open up an issue with what remains to be tested.
